presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,43 +582,193 @@ private static NestedField nestedField(String name)
             private final StandardFunctionResolution functionResolution;
             private final ExpressionOptimizer expressionOptimizer;
             private final ConnectorSession connectorSession;
-
-            private SubfieldExtractor(StandardFunctionResolution functionResolution, ExpressionOptimizer expressionOptimizer, ConnectorSession connectorSession)
+            private final FunctionAndTypeManager functionAndTypeManager;
+            private final boolean isPushDownSubfieldsFromLambdasEnabled;
+
+            private SubfieldExtractor(
+                    StandardFunctionResolution functionResolution,
+                    ExpressionOptimizer expressionOptimizer,
+                    ConnectorSession connectorSession,
+                    FunctionAndTypeManager functionAndTypeManager,
+                    boolean isPushDownSubfieldsFromLambdasEnabled)
             {
                 this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
                 this.expressionOptimizer = requireNonNull(expressionOptimizer, "expressionOptimizer is null");
-                this.connectorSession = requireNonNull(connectorSession, "connectorSession is null");
+                this.connectorSession = connectorSession;
+                this.isPushDownSubfieldsFromLambdasEnabled = isPushDownSubfieldsFromLambdasEnabled;
+                this.functionAndTypeManager = functionAndTypeManager;
             }
 
             @Override
             public Void visitCall(CallExpression call, Context context)
             {
-                if (!functionResolution.isSubscriptFunction(call.getFunctionHandle())) {
-                    call.getArguments().forEach(argument -> argument.accept(this, context));
+                ComplexTypeFunctionDescriptor functionDescriptor = functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getDescriptor();
+                if (functionResolution.isSubscriptFunction(call.getFunctionHandle()) ||
+                        functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getName().equals(ELEMENT_AT)) {
+                    Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession, functionAndTypeManager);
+                    if (subfield.isPresent()) {
+                        if (context.isPruningLambdaSubfieldsPossible()) {
```

Comment:
does this means some inner function call pruned some subfields before we got here?

Reply:
No. It means that we managed to collect the lambda subfields from all of the outer calls and we can proof that outside of the set of collected subfields, there is no other subfield required for this expression.

Reply:
Additionally, we can prune the lambda subfields only when we reached the leaf node in expression tree which could be a value reference expression, a subscript of the array or the map field, or subfield dereference.  so, it cannot be pruned by outer function call.